### PR TITLE
DataModule compatiblity with Python dataclass

### LINKED ISSUE #8272: make DataModule compatible with Python dataclass
## 🚀 Feature

Support the following:

```python
@dataclass
class MyDataModule(LightningDataModule):
    pass
```

### Motivation

To reduce boilerplate code is at the core of philosophy in Lightning. It should be compatible with dataclasses.

### Code sample

Here is an example. It currently does not work as we have some internal attributes that don't play well with the dataclass.

```python

import os
from dataclasses import dataclass

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer, LightningDataModule


class RandomDataset(Dataset):

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


@dataclass
class BoringDataModule(LightningDataModule):

    batch_size: int = 2

    def train_dataloader(self):
        return DataLoader(RandomDataset(32, 64), batch_size=self.batch_size)


class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        weights_summary=None,
    )
    trainer.fit(model, datamodule=BoringDataModule())


if __name__ == '__main__':
    run()

```

### Alternatives

#3792 introduces `save_hyperparameters()` for the datamodule. However, I believe the dataclass approach here is not in conflict with that because both could be useful at the same time.

### PR DESCRIPTION
**Pull request by himanshu-dutta**
## What does this PR do?

Adds compatibility of Python dataclass with Datamodule without breaking any currently implemented functionality.

Fixes #8272

### Does your PR introduce any breaking changes? If yes, please list them.

No.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are  #resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by himanshu-dutta** [comment]
Hey @tchaton,

Can you please review the changes and let me know if they look good to you or any changes are required?

File: tests/core/test_datamodules.py [hunk 1]
```diff
@@ -533,6 +535,81 @@ def test_simple_hyperparameters_saving():
...
+
+    trainer = Trainer(
+        default_root_dir=os.getcwd(),
+        limit_train_batches=1,
+        limit_val_batches=1,
+        num_sanity_val_steps=0,
+        max_epochs=1,
+        weights_summary=None,
+    )
+    trainer.fit(BoringModel(), datamodule=BoringDataModule2(batch_size=32))
```
Comment (awaelchli) [inline tests/core/test_datamodules.py:607]: let's not run a training here. the above assertions for the datamodule are enough to determine if it's working well as a dataclass

File: tests/core/test_datamodules.py [hunk 2]
```diff
@@ -533,6 +535,81 @@ def test_simple_hyperparameters_saving():
...
+            self.len = length
+            self.data = torch.randn(length, size)
+
+        def __getitem__(self, index):
+            return self.data[index]
+
+        def __len__(self):
+            return self.len
+
+    @dataclass(init=True, repr=True, eq=True, order=True, unsafe_hash=True, frozen=False)
```
Comment (awaelchli) [inline tests/core/test_datamodules.py:554]: these settings are not necessary, right?
we only have to add @dataclass, to enable the basic usage, right?

Comment (himanshu-dutta) [reply tests/core/test_datamodules.py:549]: Right. The first three parameters default to True, the other three to False. Only using @dataclass is enough. Have added all the parameters just to make sure, nothing breaks the functionality when enabled. Will remove for `BoringDataModule2`, and let it be for `BoringDataModule1`.

File: tests/core/test_datamodules.py (hunk 1 above)
Comment (himanshu-dutta) [reply tests/core/test_datamodules.py]: Done!

File: tests/core/test_datamodules.py [hunk 3]
```diff
@@ -533,6 +534,48 @@ def test_simple_hyperparameters_saving():
...
+
+    @dataclass
+    class BoringDataModule2(LightningDataModule):
+
+        batch_size: int = 2
+
+        def train_dataloader(self):
+            return torch.utils.data.DataLoader(RandomDataset(32, 64), batch_size=self.batch_size)
+
+    # assets for the different dunder methods added by dataclass, when super class is inherently initialized, i.e.
```
Comment (tchaton) [inline tests/core/test_datamodules.py:572]:
```suggestion
    # asserts for the different dunder methods added by dataclass, when super class is inherently initialized, i.e.
```

File: pytorch_lightning/core/datamodule.py [hunk 4]
```diff
@@ -423,6 +423,12 @@ def __new__(cls, *args: Any, **kwargs: Any) -> "LightningDataModule":
         obj.prepare_data = cls._track_data_hook_calls(obj, obj.prepare_data)
         obj.setup = cls._track_data_hook_calls(obj, obj.setup)
         obj.teardown = cls._track_data_hook_calls(obj, obj.teardown)
+
+        # calling init method of LightningDataModule when its
```
Comment (tchaton) [inline pytorch_lightning/core/datamodule.py:427]: Mind improving this comment right there, so readers who don't know abut mro understand what is happening there ?

File: tests/core/test_datamodules.py (hunk 2 above)
Comment (tchaton) [reply tests/core/test_datamodules.py:549]: Let's keep them to be sure it doesn't break. 
Anyway we can test it is actually frozen ?

**Review by tchaton — COMMENTED** [review]
Looks great ! Some small comments.

File: pytorch_lightning/core/datamodule.py (hunk 4 above)
Comment (himanshu-dutta) [reply pytorch_lightning/core/datamodule.py]: Would this be fine?

File: tests/core/test_datamodules.py (hunk 2 above)
Comment (himanshu-dutta) [reply tests/core/test_datamodules.py:549]: We can test that, but enabling frozen will break the functionality of the base class.

File: tests/core/test_datamodules.py [hunk 5]
```diff
@@ -533,6 +534,48 @@ def test_simple_hyperparameters_saving():
...
+            return self.data[index]
+
+        def __len__(self):
+            return self.len
+
+    @dataclass(init=True, repr=True, eq=True, order=True, unsafe_hash=True, frozen=False)
+    class BoringDataModule1(LightningDataModule):
+        def __init__(self, batch_size):
+            super().__init__()
+            self.batch_size = batch_size
```
Comment (carmocca) [inline tests/core/test_datamodules.py:553]: Why not use the dataclass pattern of writing attributes?

```suggestion
        batch_size: int
```

File: pytorch_lightning/core/datamodule.py [hunk 6]
```diff
@@ -423,6 +423,14 @@ def __new__(cls, *args: Any, **kwargs: Any) -> "LightningDataModule":
...
         obj.setup = cls._track_data_hook_calls(obj, obj.setup)
         obj.teardown = cls._track_data_hook_calls(obj, obj.teardown)
+
+        # mro defines the resolution order of superclasses during inheritence
+        # for a method call.
+        # We check that LightningDataModule is immediately next in line to
+        # the subclass 'cls', to make sure not to initialize any other subclass.
+        mro = cls.mro()
+        if mro[mro.index(cls) + 1] == LightningDataModule:
+            super(cls, obj).__init__()
```
Comment (carmocca) [inline pytorch_lightning/core/datamodule.py:433]: I don't understand why this `__new__` neds to call the `LightningDataModule.__init__`

Because that's what you are cheecking with that `if`, right?

File: tests/core/test_datamodules.py (hunk 5 above)
Comment (himanshu-dutta) [reply tests/core/test_datamodules.py:553]: This test is to make sure that super().__init__ call can still be made, in case the user wants to explicitly call it with parameters, from __post_init__ in a dataclass initialization. The next test makes use of the dataclass pattern, and doesn't have any __init__ or __post_init__ defined, and doesn't have a super().__init__ call either.

File: pytorch_lightning/core/datamodule.py (hunk 6 above)
Comment (himanshu-dutta) [reply pytorch_lightning/core/datamodule.py:433]: Yes. The `LightningDataModule.__init__` call initializes the private variables. If the user makes use of dataclass pattern and doesn't explicitly make the `super().__init__` call in the derived class, then we implicitly make the call in `__new__` without parameters. And in case the user does make the call somewhere (with or without the parameters) then it overrides the implicit call made in `__new__`.

`if mro[mro.index(cls) + 1] == LightningDataModule` is to make sure that incase the inheritence order is something like, `LightningDataModule->ABC->XYZ`, and `XYZ` is being initialized, we don't call `super().__init__` for it, which will point to `ABC`.

This makes sure that the users don't necessarily have to call init on LightningDataModule while using Dataclass pattern, but nothing breaks if they do.

Comment (carmocca) [reply pytorch_lightning/core/datamodule.py:433]: But if I delete this, the test added still passes.

If it's necessary, the test should check it.

File: tests/core/test_datamodules.py (hunk 2 above)
Comment (carmocca) [reply tests/core/test_datamodules.py:549]: `init` seem sunnecessary

Screenshot 2021-08-30 at 16 41 07

Comment (himanshu-dutta) [reply tests/core/test_datamodules.py:552]: Thanks for pointing the redundancy. Ideally, it should be `__post_init__`. And the other parameters are to test all the features of `dataclass` that can be enabled without breaking any functionality.

File: pytorch_lightning/core/datamodule.py (hunk 6 above)
Comment (himanshu-dutta) [reply pytorch_lightning/core/datamodule.py:433]: Added the test to check this.
Changing:
```
if mro[mro.index(cls) + 1] == LightningDataModule:
            super(cls, obj).__init__()
```
to:
```
super(cls, obj).__init__()
```
should make the `test_define_as_dataclass` fail, along with others.

Comment (carmocca) [reply pytorch_lightning/core/datamodule.py:433]: Tried again and if I remove these lines the datamodule tests still pass

Comment (himanshu-dutta) [reply pytorch_lightning/core/datamodule.py:433]: If you remove both the lines now, it would fail, because all the private variables won't be initialized.

If you just remove the `if` condition and run `super(cls, obj).__init__()`, it would fail because it wouldn't be checking for LightningDataModule as a direct parent.

File: tests/core/test_datamodules.py [hunk 7]
```diff
@@ -533,6 +534,56 @@ def test_simple_hyperparameters_saving():
     assert data.hparams == AttributeDict({"arg0": 10, "arg1": "foo", "kwarg0": "bar"})
 
 
+def test_define_as_dataclass():
+    # makes sure that no functionality is broken and the user can still manually make
+    # super().__init__ call with parameters
+    # also tests all the dataclass features that can be enabled without breaking anything
+    @dataclass(init=True, repr=True, eq=True, order=True, unsafe_hash=True, frozen=False)
+    class BoringDataModule1(LightningDataModule):
+        batch_size: int = 32
```
Comment (carmocca) [inline tests/core/test_datamodules.py:543]: I pulled the branch and tried it out with batch size as required

```suggestion
        batch_size: int
```

But it fails

File: tests/core/test_datamodules.py [hunk 8]
```diff
@@ -533,6 +534,56 @@ def test_simple_hyperparameters_saving():
...
+    # of the defined class.
+    @dataclass
+    class BoringDataModule3(SklearnDataModule):
+        num_features: int = 32
+        length: int = 800
+        num_classes: int = 3
+        batch_size: int = 2
+
+        def __post_init__(self):
+            super().__init__(
```
Comment (carmocca) [inline tests/core/test_datamodules.py:577]: For this class, `mro[mro.index(cls) + 1] == LightningDataModule` will be False correct?

Then does that mean with multiple inheritance levels, the dataclass is forced to call `__init__` in its `__post_init__`?

Comment (himanshu-dutta) [reply tests/core/test_datamodules.py:577]: Yes. The class directly inheriting from `LightningDataModule` in that case will have to make `super().__init__` call. 
 
To make it so that all the following cases pass:
Screenshot 1

there could be one solution. We can replace:
```
mro = cls.mro()
if mro[mro.index(cls) + 1] == LightningDataModule:
    super(cls, obj).__init__()
```
with just:
```
LightningDataModule.__init__(obj)
```

This would ensure that LightningDataModule is initialized, even if any further derived class is initialized or not. And can be easily overridden by a second `__init__` call.

It shouldn't break anything, would it?

File: tests/core/test_datamodules.py (hunk 7 above)
Comment (himanshu-dutta) [reply tests/core/test_datamodules.py:543]: Not sure what you mean. Removing the default value for `batch_size`wouldn't cause any problem.  If you are trying a different batch size for `BoringDataModule3`, then that's a caveat with `SklearnDataModule`.

Screenshot
And in case you are facing some problem like this, this is just how `dataclasses` create the `init` method for subclasses derived from other `dataclass`.

File: tests/core/test_datamodules.py (hunk 8 above)
Comment (carmocca) [reply tests/core/test_datamodules.py:577]: I don't think so, try pushing that! :)

Comment (himanshu-dutta) [reply tests/core/test_datamodules.py]: Have pushed the changes. Can you please allow the workflows?

Comment (himanshu-dutta) [reply tests/core/test_datamodules.py]: Works well and good. All types of multilevel inheritance cases pass, along with all the tests for direct inheritance with `dataclass`. :D

File: tests/core/test_datamodules.py [hunk 9]
```diff
@@ -533,6 +534,68 @@ def test_simple_hyperparameters_saving():
...
+    # makes sure that no functionality is broken and the user can still manually make
+    # super().__init__ call with parameters
+    # also tests all the dataclass features that can be enabled without breaking anything
+    @dataclass(init=True, repr=True, eq=True, order=True, unsafe_hash=True, frozen=False)
+    class BoringDataModule1(LightningDataModule):
+        batch_size: int
+        dims: int = 2
+
+        def __post_init__(self):
+            super().__init__(dims=self.dims)
```
Comment (awaelchli) [inline tests/core/test_datamodules.py:547]:
```suggestion
            super().__init__(dims=3)
```
don't we want to test this and assert below that dims = 3?

**Review by awaelchli — APPROVED** [review]

File: tests/core/test_datamodules.py (hunk 9 above)
Comment (himanshu-dutta) [reply tests/core/test_datamodules.py:547]: Should it matter? As long as we are asserting for the value (2 or 3) that we are overwriting through `super().__init__` call, it should be fine right?

File: pytorch_lightning/core/datamodule.py [hunk 10]
```diff
@@ -423,6 +423,10 @@ def __new__(cls, *args: Any, **kwargs: Any) -> "LightningDataModule":
         obj.prepare_data = cls._track_data_hook_calls(obj, obj.prepare_data)
         obj.setup = cls._track_data_hook_calls(obj, obj.setup)
         obj.teardown = cls._track_data_hook_calls(obj, obj.teardown)
+
+        # calling LightningDataModule.__init__ on obj, to ensure in all cases of inheritance it is called,
+        # even if explicit call to super().__init__ hasn't been made
```
Comment (carmocca) [inline pytorch_lightning/core/datamodule.py:428]: nit: we use backticks for code in comments

```suggestion
        # calling this to ensure the `LightningDataModule` is initialized for all cases of inheritance,
        # even if `super().__init__` hasn't been explicitly called in the class
```

**Review by carmocca — APPROVED** [review]
Epic!

**Review by Borda — APPROVED** [review]
